File: ios/TealiumPrismModule.swift
import Foundation
#if COCOAPODS
import TealiumPrism
#else
import TealiumPrismCore
import TealiumPrismLifecycle
#endif
import React

@objc(TealiumPrismModule)
class TealiumPrismModule: NSObject {

  private var tealium: Tealium?
  private var cachedVisitorId: String?

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func initialize(_ account: String,
                  profile: String,
                  environment: String,
                  datasource: String,
                  resolver: @escaping RCTPromiseResolveBlock,
                  rejecter: @escaping RCTPromiseRejectBlock) {

    // Release any previous instance so TealiumInstanceManager doesn't hand back a cached
    // instance keyed by account+profile — otherwise a JS-only reload would never re-fetch settingsUrl.
    tealium = nil

    let settingsUrl = "https://tags.tiqcdn.com/dle/success-robert-rizman/coffee-demo/mobile_settings_dev.json?cb=\(Int.random(in: 100000...999999))"

    let config = TealiumConfig(account: account,
                               profile: profile,
                               environment: environment,
                               dataSource: datasource,
                               modules: [],
                               settingsUrl: settingsUrl)

    tealium = Tealium.create(config: config)

    NSLog("[TealiumPrism] Initialized - account: \(account), profile: \(profile)")
    resolver(true)
  }

  private func cacheVisitorIdIfNeeded(from trackResult: TrackResult) {
    guard cachedVisitorId == nil,
          let vid: String = trackResult.dispatch.payload.get(key: TealiumDataKey.visitorId),
          !vid.isEmpty else { return }
    cachedVisitorId = vid
    NSLog("[TealiumPrism] Cached tealium_visitor_id: \(vid)")
  }

  private func dataObject(from data: NSDictionary?) -> DataObject {
    guard let dict = data as? [String: Any],
          let dataObject = try? DataObject(jsonObject: dict) else {
      return DataObject()
    }
    return dataObject
  }

  @objc
  func track(_ eventName: String,
             data: NSDictionary,
             resolver: @escaping RCTPromiseResolveBlock,
             rejecter: @escaping RCTPromiseRejectBlock) {

    guard let tealium = tealium else {
      rejecter("NOT_INITIALIZED", "Tealium not initialized", nil)
      return
    }

    let payload = dataObject(from: data)

    tealium.track(eventName, type: .event, data: payload).onSuccess { [weak self] trackResult in
      self?.cacheVisitorIdIfNeeded(from: trackResult)
    }

    resolver(true)
  }

  @objc
  func trackView(_ screenName: String,
                 data: NSDictionary,
                 resolver: @escaping RCTPromiseResolveBlock,
                 rejecter: @escaping RCTPromiseRejectBlock) {

    guard let tealium = tealium else {
      rejecter("NOT_INITIALIZED", "Tealium not initialized", nil)
      return
    }

    var payload = dataObject(from: data)
    payload.set(screenName, key: "screen_name")

    tealium.track(screenName, type: .view, data: payload).onSuccess { [weak self] trackResult in
      self?.cacheVisitorIdIfNeeded(from: trackResult)
    }

    resolver(true)
  }

  @objc
  func setDataLayer(_ data: NSDictionary,
                    resolver: @escaping RCTPromiseResolveBlock,
                    rejecter: @escaping RCTPromiseRejectBlock) {

    guard let tealium = tealium else {
      rejecter("NOT_INITIALIZED", "Tealium not initialized", nil)
      return
    }

    tealium.dataLayer.put(data: dataObject(from: data))

    resolver(true)
  }

  @objc
  func joinTrace(_ traceId: String,
                 resolver: @escaping RCTPromiseResolveBlock,
                 rejecter: @escaping RCTPromiseRejectBlock) {

    guard let tealium = tealium else {
      rejecter("NOT_INITIALIZED", "Tealium not initialized", nil)
      return
    }

    tealium.trace.join(id: traceId)
    NSLog("[TealiumPrism] Joined trace: \(traceId)")
    resolver(true)
  }

  @objc
  func leaveTrace(_ resolver: @escaping RCTPromiseResolveBlock,
                  rejecter: @escaping RCTPromiseRejectBlock) {

    guard let tealium = tealium else {
      resolver(false)
      return
    }

    tealium.trace.leave()
    NSLog("[TealiumPrism] Left trace")
    resolver(true)
  }

  @objc
  func getAppUuid(_ resolver: @escaping RCTPromiseResolveBlock,
                  rejecter: @escaping RCTPromiseRejectBlock) {

    guard let tealium = tealium else {
      rejecter("NOT_INITIALIZED", "Tealium not initialized", nil)
      return
    }

    tealium.dataLayer.get(key: TealiumDataKey.appUUID, as: String.self).onSuccess { uuid in
      if let uuid, !uuid.isEmpty {
        resolver(uuid.lowercased())
      } else {
        rejecter("NO_UUID", "app_uuid not available", nil)
      }
    }
  }

  @objc
  func getVisitorId(_ resolver: @escaping RCTPromiseResolveBlock,
                    rejecter: @escaping RCTPromiseRejectBlock) {

    if let vid = cachedVisitorId, !vid.isEmpty {
      NSLog("[TealiumPrism] getVisitorId: \(vid)")
      resolver(vid)
    } else {
      rejecter("NO_VISITOR_ID", "tealium_visitor_id not available", nil)
    }
  }
}

File: android/app/src/main/java/com/tealium/coffeecafe/TealiumPrismModule.kt
package com.tealium.coffeecafe

import android.app.Application
import android.util.Log
import com.facebook.react.bridge.*
import com.tealium.prism.core.api.Tealium
import com.tealium.prism.core.api.TealiumConfig
import com.tealium.prism.core.api.data.DataObject
import com.tealium.prism.core.api.logger.LogHandler
import com.tealium.prism.core.api.logger.LogLevel

class TealiumPrismModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "TealiumPrismModule"

    companion object {
        private var tealium: Tealium? = null
        private var cachedVisitorId: String? = null
    }

    private fun cacheVisitorIdIfNeeded(result: com.tealium.prism.core.api.misc.TealiumResult<com.tealium.prism.core.api.tracking.TrackResult>) {
        if (cachedVisitorId != null) return
        result.getOrNull()?.dispatch?.payload()?.getString("tealium_visitor_id")?.let { vid ->
            if (vid.isNotEmpty()) {
                cachedVisitorId = vid
                println("[TealiumPrism] Cached tealium_visitor_id: $vid")
            }
        }
    }

    @ReactMethod
    fun initialize(account: String, profile: String, environment: String, datasource: String, promise: Promise) {
        try {
            // Unlike iOS's weak-referenced instance registry (where releasing our reference lets
            // the old instance deallocate automatically), Android's TealiumInstanceManager holds
            // instances in a plain MutableMap with strong references — Tealium.create(config) will
            // keep handing back the original instance forever, ignoring any new settingsUrl, unless
            // we explicitly shut it down first.
            tealium = null
            Tealium.shutdown("$account-$profile")

            val app = reactApplicationContext.applicationContext as Application
            val settingsUrl = "https://tags.tiqcdn.com/dle/success-robert-rizman/coffee-demo/mobile_settings_dev.json?cb=${(100000..999999).random()}"
            val config = TealiumConfig.Builder(
                application = app,
                accountName = account,
                profileName = profile,
                environment = environment,
                modules = listOf()
            ).setDataSource(datasource)
                .setSettingsUrl(settingsUrl)
                .configureCoreSettings { it.setLogLevel(LogLevel.TRACE) }
                .setLogHandler(object : LogHandler {
                    override fun log(tag: String, message: String, level: LogLevel) {
                        Log.d("TealiumPrismSDK", "[${level.name}] $tag: $message")
                    }
                })
                .build()

            tealium = Tealium.create(config)
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("INIT_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun track(eventName: String, data: ReadableMap?, promise: Promise) {
        val t = tealium ?: run { promise.reject("NOT_INITIALIZED", "PRISM not initialized"); return }
        try {
            t.track(eventName, buildDataObject(data)).subscribe { result ->
                if (result.isFailure) {
                    Log.e("TealiumPrismSDK", "track('$eventName') failed", result.exceptionOrNull())
                }
                cacheVisitorIdIfNeeded(result)
            }
            promise.resolve(true)
        } catch (e: Exception) { promise.reject("TRACK_ERROR", e.message, e) }
    }

    @ReactMethod
    fun trackView(screenName: String, data: ReadableMap?, promise: Promise) {
        val t = tealium ?: run { promise.reject("NOT_INITIALIZED", "PRISM not initialized"); return }
        try {
            t.track(screenName, buildDataObject(data)).subscribe { result ->
                if (result.isFailure) {
                    Log.e("TealiumPrismSDK", "trackView('$screenName') failed", result.exceptionOrNull())
                }
                cacheVisitorIdIfNeeded(result)
            }
            promise.resolve(true)
        } catch (e: Exception) { promise.reject("TRACK_ERROR", e.message, e) }
    }

    @ReactMethod
    fun setDataLayer(data: ReadableMap, promise: Promise) {
        val t = tealium ?: run { promise.reject("NOT_INITIALIZED", "PRISM not initialized"); return }
        try {
            val iter = data.keySetIterator()
            while (iter.hasNextKey()) {
                val key = iter.nextKey()
                when (data.getType(key)) {
                    ReadableType.String -> t.dataLayer.put(key, data.getString(key) ?: "")
                    ReadableType.Number -> t.dataLayer.put(key, data.getDouble(key))
                    ReadableType.Boolean -> t.dataLayer.put(key, data.getBoolean(key))
                    else -> {}
                }
            }
            promise.resolve(true)
        } catch (e: Exception) { promise.reject("DATALAYER_ERROR", e.message, e) }
    }

    @ReactMethod
    fun joinTrace(traceId: String, promise: Promise) {
        val t = tealium ?: run { promise.reject("NOT_INITIALIZED", "PRISM not initialized"); return }
        try { t.trace.join(traceId); promise.resolve(true) }
        catch (e: Exception) { promise.reject("TRACE_ERROR", e.message, e) }
    }

    @ReactMethod
    fun leaveTrace(promise: Promise) {
        try { tealium?.trace?.leave(); promise.resolve(true) }
        catch (e: Exception) { promise.reject("TRACE_ERROR", e.message, e) }
    }

    private fun buildDataObject(map: ReadableMap?): DataObject {
        return DataObject.create {
            map?.keySetIterator()?.let { iter ->
                while (iter.hasNextKey()) {
                    val key = iter.nextKey()
                    when (map.getType(key)) {
                        ReadableType.String -> put(key, map.getString(key) ?: "")
                        ReadableType.Number -> put(key, map.getDouble(key))
                        ReadableType.Boolean -> put(key, map.getBoolean(key))
                        else -> {}
                    }
                }
            }
        }
    }

    @ReactMethod
    fun getAppUuid(promise: Promise) {
        val t = tealium ?: run { promise.reject("NOT_INITIALIZED", "PRISM not initialized"); return }
        try {
            t.dataLayer.getString("app_uuid").subscribe { result ->
                val uuid = result.getOrNull()
                if (!uuid.isNullOrEmpty()) {
                    println("[TealiumPrism] app_uuid: $uuid")
                    promise.resolve(uuid.lowercase())
                } else {
                    promise.reject("NO_UUID", "app_uuid not available")
                }
            }
        } catch (e: Exception) {
            promise.reject("UUID_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun getVisitorId(promise: Promise) {
        // Return the value intercepted from the enriched track() payload
        val cached = cachedVisitorId
        if (!cached.isNullOrEmpty()) {
            println("[TealiumPrism] getVisitorId from cache: $cached")
            promise.resolve(cached)
            return
        }
        // Not cached yet — events haven't fired or the dispatch payload didn't contain it
        promise.reject("NO_VISITOR_ID", "tealium_visitor_id not yet available")
    }
}